build-system/tasks/e2e/validate-tests.js

```diff
@@ -0,0 +1,302 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific lan``guage governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const argv = require('minimist')(process.argv.slice(2));
+const testOverrides = require('../../test-configs/test-overrides.json');
+const {cyan, green, red} = require('../../common/colors');
+const {getDefaultMochaFiles} = require('./mocha-utils');
+const {getFilesFromArgv} = require('../../common/utils');
+const {gitDiffNameOnlyMain} = require('../../common/git');
+const {log} = require('../../common/logging');
+const {Worker} = require('worker_threads');
+
+/**
+ * @typedef {{
+ *   startTime: number,
+ *   endTime: number,
+ *   passed: boolean,
+ * }}
+ */
+let TestResultDef;
+
+/**
+ * Get the maximum allowed average runtime of a test.
+ * @return {number}
+ */
+function getMaxAllowedTestTime_() {
+  return process.env.MAX_TEST_TIME
+    ? parseInt(process.env.MAX_TEST_TIME, 10)
+    : 1000;
+}
+
+/**
+ * Get the maximum allowed average runtime for a "LARGE" test.
+ * Tests are defined as "LARGE" in the build-system/test-configs/test-overrides.json file.
+ * @return {number}
+ */
+function getMaxLargeTestTime_() {
+  return process.env.MAX_LARGE_TEST_TIME
+    ? parseInt(process.env.MAX_LARGE_TEST_TIME, 10)
+    : 5000;
+}
+
+/**
+ * Get the maximum allowed average runtime for a "XLARGE" test.
+ * Tests are defined as "XLARGE" in the build-system/test-configs/test-overrides.json file.
+ * @return {number}
+ */
+function getMaxXLargeTestTime_() {
+  return process.env.MAX_XLARGE_TEST_TIME
+    ? parseInt(process.env.MAX_XLARGE_TEST_TIME, 10)
+    : 10000;
+}
```

Comment:
Allowable test runtimes can all be overridden with environment variables if something changes in CI and they need to be altered quickly to get jobs passing again or for running locally on a slower machine.

```
normal: 1s
large: 5s
x-large: 10s
```

Reply:
Note that tests will time out by default after `3s` unless their timeout is set now that #35273 has been submitted